File: python/vedadb_orm/cache.py
"""
VedaDB ORM — CacheProxy.

Thin convenience layer on top of VedaDB's native CACHE commands, scoped
to a model class.
"""

from __future__ import annotations

import json
from typing import Any, Optional, TYPE_CHECKING

from .exceptions import QueryError

if TYPE_CHECKING:
    from .query import QueryBuilder


class CacheProxy:
    """Model-scoped cache operations using VedaDB's CACHE engine."""

    def __init__(self, model_cls: type) -> None:
        self._model_cls = model_cls
        self._prefix = model_cls._table_name

    def _orm(self) -> Any:
        orm = self._model_cls._get_orm()
        if orm is None:
            raise QueryError("Model is not bound to a VedaORM instance")
        return orm

    def _key(self, key: str) -> str:
        return f"{self._prefix}:{key}"

    def get(self, key: str) -> Any:
        """Retrieve a cached value by key."""
        result = self._orm().query(f"CACHE GET '{self._key(key)}';")
        if result.rows and result.rows[0]:
            raw = result.rows[0][0]
            try:
                return json.loads(raw)
            except (json.JSONDecodeError, TypeError):
                return raw
        return None

    def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Store a value in the cache."""
        if isinstance(value, (dict, list)):
            formatted = json.dumps(value)
        else:
            formatted = f"'{value}'"
        sql = f"CACHE SET '{self._key(key)}' {formatted}"
        if ttl is not None:
            sql += f" TTL {ttl}"
        self._orm().query(sql + ";")

    def delete(self, key: str) -> None:
        """Remove a cached key."""
        self._orm().query(f"CACHE DEL '{self._key(key)}';")

    def invalidate(self, **where: Any) -> None:
        """Invalidate cache entries whose key contains the given predicates.

        A simple convention: keys are stored as ``table:field=value``, so
        this builds the matching key suffix and deletes it.
        """
        parts = [f"{k}={v}" for k, v in where.items()]
        key = ":".join(parts)
        self.delete(key)

    def wrap_query(self, query_builder: "QueryBuilder", ttl: Optional[int] = None) -> Any:
        """Cache-aside pattern: check cache first, fall back to query.

        The cache key is derived from the generated SQL.
        """
        import hashlib
        sql = query_builder.to_sql()
        cache_key = "qry:" + hashlib.md5(sql.encode()).hexdigest()

        cached = self.get(cache_key)
        if cached is not None:
            return cached

        results = query_builder.all()
        serialized = [inst.to_dict() for inst in results]

        effective_ttl = ttl
        if effective_ttl is None:
            schema = getattr(self._model_cls, "__schema__", None)
            if schema and schema.cache.enabled:
                effective_ttl = schema.cache.ttl

        self.set(cache_key, serialized, effective_ttl)
        return results

File: node/vedadb-orm/src/cache.ts
/**
 * VedaDB ORM CacheProxy
 *
 * Provides model-level caching via the VedaDB CACHE sub-API.
 */

import { DriverClient, DriverResult } from './types';
import { SchemaDefinition } from './schema';
import { escapeValue } from './utils';

export class CacheProxy<T = any> {
  private schema: SchemaDefinition<T>;
  private executor: (sql: string) => Promise<DriverResult>;
  private defaultTTL: number;

  constructor(
    schema: SchemaDefinition<T>,
    executor: (sql: string) => Promise<DriverResult>,
    defaultTTL: number = 300
  ) {
    this.schema = schema;
    this.executor = executor;
    this.defaultTTL = defaultTTL;
  }

  /**
   * Build a cache key for a given model record by its primary key.
   */
  private cacheKey(id: any): string {
    return `${this.schema.tableName}:${id}`;
  }

  /**
   * Store a model instance in cache.
   */
  async set(id: any, data: Partial<T>, ttl?: number): Promise<void> {
    const key = this.cacheKey(id);
    const value = JSON.stringify(data).replace(/'/g, "''");
    const t = ttl ?? this.defaultTTL;
    await this.executor(`CACHE SET '${key}' '${value}' TTL ${t};`);
  }

  /**
   * Retrieve a model instance from cache.
   */
  async get(id: any): Promise<Partial<T> | null> {
    const key = this.cacheKey(id);
    const result = await this.executor(`CACHE GET '${key}';`);
    const obj = result.first();
    if (!obj) return null;
    const raw = obj['value'] || obj[Object.keys(obj)[0]];
    if (!raw) return null;
    try {
      return JSON.parse(String(raw)) as Partial<T>;
    } catch {
      return null;
    }
  }

  /**
   * Delete a model instance from cache.
   */
  async del(id: any): Promise<void> {
    const key = this.cacheKey(id);
    await this.executor(`CACHE DEL '${key}';`);
  }

  /**
   * Invalidate all cache entries for this model.
   */
  async invalidateAll(): Promise<void> {
    const pattern = `${this.schema.tableName}:*`;
    await this.executor(`CACHE DEL '${pattern}';`);
  }

  /**
   * Get-or-set: return cached value or execute finder and cache result.
   */
  async getOrSet(id: any, finder: () => Promise<T | null>, ttl?: number): Promise<T | null> {
    const cached = await this.get(id);
    if (cached) return cached as T;

    const value = await finder();
    if (value) {
      await this.set(id, value as Partial<T>, ttl);
    }
    return value;
  }
}
